code/modules/atmospherics/machinery/pipes/smart.dm

```diff
@@ -0,0 +1,114 @@
+/obj/machinery/atmospherics/pipe/smart
+	icon = 'icons/obj/atmospherics/pipes/simple.dmi'
+	icon_state = "pipe11-3"
+
+	name = "pipe"
+	desc = "A one meter section of regular pipe."
+
+	vis_flags = VIS_INHERIT_ICON | VIS_INHERIT_ICON_STATE | VIS_INHERIT_DIR | VIS_INHERIT_ID
+
+	device_type = QUATERNARY
+	construction_type = /obj/item/pipe/quaternary
+	pipe_state = "manifold4w"
+	var/connection_num = 0
+	var/list/connections
+	var/static/list/mutable_appearance/center_cache = list()
+	var/mutable_appearance/pipe_appearance
+
+/obj/machinery/atmospherics/pipe/smart/New()
+	icon_state = ""
+	connections = new/list(dir2text(NORTH) = FALSE, dir2text(SOUTH) = FALSE , dir2text(EAST) = FALSE , dir2text(WEST) = FALSE)
+	return ..()
+
+/obj/machinery/atmospherics/pipe/smart/Initialize()
+	. = ..()
+	update_appearance()
+
+/obj/machinery/atmospherics/pipe/smart/SetInitDirections()
+	initialize_directions = ALL_CARDINALS
+
+/obj/machinery/atmospherics/pipe/smart/proc/check_connections()
+	var/mutable_appearance/center
+	connection_num = 0
+	connections = list(dir2text(NORTH) = FALSE, dir2text(SOUTH) = FALSE , dir2text(EAST) = FALSE , dir2text(WEST) = FALSE)
```

Comment:
I think a bitfield for this would be better than an assoc list, since all the directionals are bitflags anyway. Would likely make some other stuff nicer.

Reply:
I tried to make this a bitfield, but the directions are weird and i cannot manage to make it work properly, mostly because the pipes need a very precise direction check and with the bitfield it seems to be dependable on the order you build the pipes

Reply:
The moth is right here, we can figure out the issue more in coderbus but if the issue is just in being able to iterate over the direction, that's what `GLOB.cardinals` is for

Or the slightly too clever
```
for(var/i in 0 to 3)
  var/dir = 1 << i
```

Reply:
done, used flag instead of the dir2text mess